/*
 * Copyright(c) 2012-2018 Intel Corporation
 * SPDX-License-Identifier: BSD-3-Clause-Clear
 */
syntax = "proto3";
option cc_generic_services = true;
package octf.proto.trace;

/**
 * @brief Generic trace event
 */
message EventHeader {
    /** Event sequence ID */
    uint64 sid = 1;

    /** Time stamp */
    uint64 timestamp = 2;
}

/**
 *  @brief Device description event - provides mapping between
 *  device id and name
 */
message EventDeviceDescription {
    /** Device Id */
    uint64 id = 1;

    /** Device Name */
    string name = 2;

    /** Device Size */
    uint32 size = 3;
}

/** @brief IO operation direction */
enum IoType {
    Read = 0;
    Write = 1;
    Discard = 2;
}

/**
 * @brief IO trace event
 */
message EventIo {
    /** Address of IO in sectors */
    uint64 lba = 1;

    /** Size of IO in sectors */
    uint32 len = 2;

    /** IO class of IO */
    uint32 ioClass = 3;

    /** Device ID */
    uint64 deviceId = 4;

    /** Operation type: read, write, trim */
    IoType operation = 5;

    /** Flush flag */
    bool flush = 6;

    /** FUA flag */
    bool fua = 7;
}

message Event {
    /** Trace event header */
    EventHeader header = 1;

    oneof EventType {
        EventIo io = 2;
        EventDeviceDescription deviceDescription = 3;
    }
}
